### PR TITLE
Add a client instance to function handler params

### PR DESCRIPTION
**Pull request by selfcontained**
###  Summary

**Version URL:**
```
https://raw.githubusercontent.com/slackapi/deno-slack-sdk/32d177e74c5c9c54885d7572dd52c20f9d55893c/src/mod.ts
```

This extends the context argument passed into function handlers (main handler and interactivity handlers) to include a `client` property, which is a `SlackAPI()` instance on behalf of the `deno-slack-api` library.

Function handlers can make Slack API calls w/ the provided `client` property.
```ts
export default SlackFunction(MyFunction, async ({ inputs, client }) => {
    const resp = await client.chat.postMessage({
      channel: inputs.approval_channel_id,
      blocks: {
        //...
      },
    });

    if (!resp.ok) {
      console.log("Error posting message", resp);
    }

    return {
      completed: false,
    };
  },
)
```

Additional handlers, such as `blockActions`, `viewSubmission`, `viewClosed` and the `unhandledError` handler also receive a `client` instance.

### Details
* The client instance is created with the `token` that the [deno-slack-runtime](https://github.com/slackapi/deno-slack-runtime) provides all handlers.
* If `env["SLACK_API_URL"]` is set, the client will set the `slackApiUrl` configuration option with it.
* `deno-slack-api` becomes a dependency of this library, so we'll want to keep it current as we publish new versions of it.

### Questions I'd love input/opinions on
* Since the `deno-slack-api` becomes an internal dep of this library, should we include some top level exports of it for convenience? If so, what exports do you think we should include?

### Testing
I've updated and added some unit tests to ensure a `client` object is present in each of the 4 handler types.

You can also manually test this by updating your `deno-slack-sdk` import map to the following (you can also drop the api entry once you've swapped over to the client in the handler args):

```
{
  "imports": {
    "deno-slack-sdk/": "https://raw.githubusercontent.com/slackapi/deno-slack-sdk/32d177e74c5c9c54885d7572dd52c20f9d55893c/src/"
  }
}

```

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/{project_slug}/blob/main/.github/CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

File: build_npm.ts [hunk 1]
```diff
@@ -27,4 +27,4 @@ await build({
 });
 
 // post build steps
-Deno.copyFileSync("README.md", "npm/README.md");
\ No newline at end of file
+Deno.copyFileSync("README.md", "npm/README.md");
```
Comment (selfcontained) [inline build_npm.ts:30]: new version of deno has some `fmt` adjustments that this picked up.

**Review by filmaj — CHANGES_REQUESTED** [review]
Looks good but doesn't seem to work with the old `SlackFunctionHandler` approach.

Other than that, I think keeping the existing export name for the slackAPI as the main deno-slack-api module does (`SlackAPI`) makes sense.

One more note: it looks like with `enrichContext` we are inching towards a middleware-like pattern. Perhaps something to keep in mind in case we want to add further enrichment of different kinds down the road. Worth noting that all of the various languages that Bolt supports implement the middleware pattern. This only applies to the old platform, but, apps implementing OAuth (to either enable distribution on the Slack app store or to assume permissions on behalf of a user to do things like post messages as a user or update a user's status) would regularly implement a middleware layer that would look up installation information (like team id and interacting user id) to retrieve the right token to use for specific events (particularly necessary for apps that are installed to multiple unrelated Workspaces). This was such a common pattern that we implemented '[Installation Stores](https://slack.dev/bolt-js/concepts#authenticating-oauth)' in Bolt. While not needed for deno-slack-sdk apps, I think it's a salient use case to keep in mind moving forward. I could see devs wanting to, down the road, use a similar middleware-layer pattern to do things like retrieve user-specific contextual information from their own datastores to inject directly into handler contexts.

It wouldn't surprise me if 6-12 months down the road we refactor the approach implemented in this PR to more generally implement middleware.

**Review by shapirone — COMMENTED** [review]
Dropping some comments since I know we're revisiting the problem Fil mentioned above

File: src/functions/enrich-context.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  BaseRuntimeFunctionContext,
+  FunctionContextEnrichment,
+} from "./types.ts";
+
+export const enrichContext = (
+  // deno-lint-ignore no-explicit-any
+  context: BaseRuntimeFunctionContext<any>,
+): typeof context & FunctionContextEnrichment => {
+  const token = context.token;
+  const slackApiUrl = (context.env || {})["SLACK_API_URL"];
```
[the remark this answers is not in the document]
Comment (selfcontained) [reply src/functions/enrich-context.ts:12]: At runtime this is reliant on w/e `deno-slack-runtime` passes in to the top level handler. While it "should" be set, and is currently, this is a safety measure against something changing on accident in `deno-slack-runtime`.

File: src/functions/slack-function.ts [hunk 3]
```diff
@@ -21,11 +27,21 @@ export const SlackFunction = <
...
-  const handlerModule: any = (...args) => functionHandler(...args);
+  const handlerModule: any = (
+    ctx: RuntimeFunctionContext<InputParameters>,
+    // deno-lint-ignore no-explicit-any
+    ...args: any
+  ) => {
+    // enrich the context w/ additional properties
+    const newContext = enrichContext(ctx);
+
+    //@ts-ignore - intentionally specifying the provided functionHandler as the `this` arg for the handler's call
```
[the remark this answers is not in the document]
Comment (filmaj) [reply src/functions/slack-function.ts]: LMAO that @ts-expect-error and @ts-ignore are actual GitHub users 😂

File: docs/functions-action-handlers.md [hunk 4]
```diff
@@ -89,49 +89,44 @@ subdirectory in your app) that posts a message with two buttons: an approval but
...
-        "block_id": "mah-buttons",
-        "elements": [{
-          type: "button",
-          text: {
-            type: "plain_text",
-            text: "Approve",
-          },
-          action_id: "approve_request",
-          style: "primary",
+export default SlackFunction(ApprovalFunction, async ({ inputs, client }) => {
```
Comment (filmaj) [inline docs/functions-action-handlers.md:96]: Oh man good call to update these docs - totally slipped my mind!

**Review by filmaj — APPROVED** [review]
Works with the ol' `SlackFunctionHandler`!

File: src/functions/slack-function.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (selfcontained) [reply src/functions/slack-function.ts]: Love the hover cards 😄

**Pull request merged**